felix/dataplane/linux/bpf_ep_mgr.go

```diff
@@ -2968,11 +2968,11 @@ func (m *bpfEndpointManager) extractTiers(tiers []*proto.TierInfo, direction Pol
 			}
 
 			for i, polName := range directionalPols {
-				staged := model.PolicyIsStaged(polName)
-
-				if !staged {
-					stagedOnly = false
+				if model.PolicyIsStaged(polName) {
+					logrus.Debugf("SKipping staged policy %v", polName)
+					continue
 				}
+				stagedOnly = false
```

Comment:
why is this always false? To make sure that if there is a tier with staged only policies, we still honor that tier somehow? Do we need to test for this case somehow? Are we testing it already?

Reply:
If we reach this point, it means we are processing a none staged policy. So, it means this tier is not one with only staged only policies. As such `stagedOnly` is set to false.

This is similar to what it was before. The only change is to skip processing/programming staged policies.

Reply:
The difference is that if we are handling a tier with only staged policies, the end of tier action would be different: 
```
if endTierDrop && !stagedOnly && tier.DefaultAction != string(apiv3.Pass) {
				polTier.EndRuleID = m.endOfTierDropID(dir, tier.Name)
				polTier.EndAction = polprog.TierEndDeny
			} else {
				polTier.EndRuleID = m.endOfTierPassID(dir, tier.Name)
				polTier.EndAction = polprog.TierEndPass
			}
```

Reply:
We have a FV for this: https://github.com/projectcalico/calico/blob/3185239e17c50f85d54d48a2abbdb37f60d0968c/felix/fv/tiered_policy_test.go#L377

Reply:
btw, this change was made with introducing tier's default action: https://github.com/projectcalico/calico/pull/9232

Reply:
My question is that you set it to true, then you skip all staged policies to set it to false.

Shouldn't that for loop be like `for policies { if policy is not staged { stagedOnly = false ; break loop ; } }` because what you are looking for it that there is at least one non-staged policy, right? You write the kinda same thing, but for a tier with only non-staged policies you loop over all of them while you can bail out much sooner. This way it reads weird and perhaps therefore I am reading it wrong :thinking:

Reply:
The loop performs two things:
1. Detecting if there is any non-staged policy in tier.
2. Programming non-staged policies.

If we break out of the loop when encountering non-staged policy, then we won't program non-staged policies.
The confusion is due to the fact that the loop is trying to perform two things in one loop.

Reply:
Ohhhh I get it now, sorry, yeah, all good! I misread where you get out of the loop :heavy_check_mark:

Reply:
To make it more readable we need to break it into two loop, but it would be less efficient.